cumulus/client/consensus/aura/src/collators/lookahead.rs

```diff
@@ -214,19 +213,28 @@ where
 				},
 			};
 
+			let slot_duration = match sc_consensus_aura::slot_duration(&*params.para_client) {
```

Comment:
```suggestion
			let slot_duration = match sc_consensus_aura::slot_duration_at(&*params.para_client) {
```

You should use this. This means that you will need to move all this down until we have the actual parent we want to build up on.

Reply:
Isn't it a chicken-and-egg situation? We need the slot duration for `can_build_upon()`, which is needed to find the parent we're building upon, which is needed to acquire the slot duration?

Reply:
Tried the best I could, so I just took the `initial_parent` we're building upon and used it as a pivot point to request `slot_duration_at()`. Got the same error as before. To me, it seems quite logical, as when that initial parent was built, the slot duration was 12s, so that's what we actually see here. But maybe I don't understand something you have in mind.

Reply:
```
			let can_build_upon = |block_hash| {
                                			let (slot_now, timestamp) = match consensus_common::relay_slot_and_timestamp(
				&relay_parent_header,
				slot_duration_at(block_hash),
			) {
				None => continue,
				Some((r_s, t)) => {
					let our_slot = Slot::from_timestamp(t, params.slot_duration);
					tracing::debug!(
						target: crate::LOG_TARGET,
						relay_slot = ?r_s,
						para_slot = ?our_slot,
						timestamp = ?t,
						slot_duration = ?params.slot_duration,
						relay_chain_slot_duration = ?params.relay_chain_slot_duration,
						"Adjusted relay-chain slot to parachain slot"
					);
					(our_slot, t)
				},
			};


				can_build_upon::<_, _, P>(
					slot_now,
					timestamp,
					block_hash,
					included_block,
					para_client,
					&keystore,
				)
			};
```

Shouldn't something like this work? I mean yes, we would need to propagate the `continue` out of `can_build_upon`. However, recomputing the `slot_duration` sounds sensible to me?

Reply:
> ```
> 			let can_build_upon = |block_hash| {
>           			let (slot_now, timestamp) = match consensus_common::relay_slot_and_timestamp(
> 					&relay_parent_header,
> 					slot_duration_at(block_hash),
> 			) {
> ```

I believe the second argument must be the relay chain slot duration, not the parachain one?
In that case, I don't see how it is different from just using parachain slot duration at `initial_parent.hash` (that's what would be passed to `can_build_upon()` in the first place), but I'll give it a try anyway.

Reply:
My point is that you can only request slot duration at some *previous* block. And for any previous block it's guaranteed to be 12s.

Reply:
Okay, I tried it. Good news for you: it works. Bad news for me: now I have to understand why it works :thinking: